compiler/llvm/clangUtil.cpp

```diff
@@ -602,171 +579,545 @@ static void handleMacroTokens(const MacroInfo* inMacro,
+      }
+    } else {
+      IF1_float_type size = FLOAT_SIZE_64;
+
+      if (tolower(numString[numString.length() - 1]) == 'l') {
+        numString[numString.length() - 1] = '\0';
+        // TODO: use long double support once it exists
+      } else if (tolower(numString[numString.length() - 1]) == 'f') {
+        numString[numString.length() - 1] = '\0';
+        size = FLOAT_SIZE_32;
+      }
+
+      varRet = new_RealSymbol(numString.c_str(), size);
     }
-    case tok::string_literal: {
-      std::string body = std::string(tok.getLiteralData(), tok.getLength());
-      if( debugPrint) printf("str = %s\n", body.c_str());
-      varRet = new_CStringSymbol(body.c_str());
-      break;
+
+    INT_ASSERT(varRet != NULL);
+    *imm = *varRet->immediate;
+    return true;
+  }
+
+  return false;
+}
+
+static bool handleNumericBinOpExpr(const MacroInfo* inMacro,
+                                   MacroInfo::tokens_iterator start,
+                                   MacroInfo::tokens_iterator end,
+                                   Immediate* imm) {
+  // handle select binary operators
+  // this only works if the LHS and RHS are either:
+  //  parenthesized expressions; or
+  //  numeric constants
+  bool lhsOk = false;
+  MacroInfo::tokens_iterator lhsStart = start;
+  MacroInfo::tokens_iterator lhsEnd = start;
+
+  // find a LHS constant or parenthesized-expression
+  if (start->getKind() == tok::numeric_constant) {
+    lhsOk = true;
+    lhsStart = start;
+    lhsEnd = start+1;
+  } else {
+    lhsOk = findParenthesizedExpr(inMacro, start, end, lhsStart, lhsEnd);
+  }
+
+  if (lhsOk == false || lhsEnd == end || lhsEnd+1 == end )
+    return false;
+
+  bool rhsOk = false;
+  MacroInfo::tokens_iterator op = lhsEnd;
+  MacroInfo::tokens_iterator rhsStart = op + 1;
+  MacroInfo::tokens_iterator rhsEnd = rhsStart;
+
+  // find a RHS constant or parenthesized-expression
+  if (rhsStart->getKind() == tok::numeric_constant) {
+    rhsOk = true;
+    rhsEnd = rhsStart+1;
+  } else {
+    rhsOk = findParenthesizedExpr(inMacro, rhsStart, end, rhsStart, rhsEnd);
+  }
+
+  if (rhsOk == false)
+    return false;
+
+  // fail if it didn't cover everything
+  if (lhsStart != start || rhsEnd != end)
+    return false;
+
+  // Compute the lhs and rhs immediates
+  Immediate lhsImm;
+  Immediate rhsImm;
+  const char* lhsCastToTy = NULL;
+  const char* rhsCastToTy = NULL;
+
+  lhsOk = handleNumericExpr(inMacro, lhsStart, lhsEnd, &lhsImm, lhsCastToTy);
+  rhsOk = handleNumericExpr(inMacro, rhsStart, rhsEnd, &rhsImm, rhsCastToTy);
+  if (lhsOk == false || rhsOk == false)
+    return false;
+  if (lhsCastToTy != NULL || rhsCastToTy != NULL)
+    return false;
+
+  // Apply the binary operator to the immediate
+  if (op->getKind() == tok::lessless) {
+    int p = 0;
+    switch (op->getKind()) {
+      case tok::lessless:   p = P_prim_lsh;    break;
+      default:
+        INT_FATAL("unhandled case");
```

Comment:
I have questions here:

- You have alluded to this in the PR message but why only `<<`? Is it only because driven by a use case, or are there fundamental issues with other simple binary operators? It is a bit awkward to support left shift but not the right shift
- Why do we have a conditional and a switch both on `op->getKind`?

Reply:
I didn't want to bother to implement them all right now.  Indeed, `<<` is driven by a specific use case.
I've updated to just have a switch.